File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        start = 0
        result = 0
        counter = 0
        char_set = set()
        for i in range(len(s)):
            ch = s[i]
            while ch in char_set:
                char_set.remove(s[start])
                start = start + 1
            
            char_set.add(ch)
            result = max(result, i - start + 1)
        return result

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int counter = 0;
        int result = 0;
        int start = 0;
        Set<Character> sett = new HashSet<>();
        for(int i = 0; i < s.length(); i++){
            char ch = s.charAt(i);
            if(sett.contains(ch)){
                while(sett.contains(ch)){
                    sett.remove(s.charAt(start));
                    start++;
                }
                sett.add(ch);
                counter = i-start+1;
            } else {
                sett.add(ch);
                counter++;
            }
            result = Math.max(result, counter);
        }
        return result;
        
    }
}